_posts/2019-05-10-server-rendering-responsively.md

```diff
@@ -0,0 +1,189 @@
+component sub-trees. We couldn't find an approach that satisfied our needs, so we wrote our own.
+
+## Tool 2: [`@artsy/react-responsive-media`](https://github.com/artsy/react-responsive-media)
+
+First off, an announcement: we've just released `react-responsive-media` version 1.0! (TODO: release v1.0 😬 or
+change this announcement)
+
+`react-responsive-media` allows you to define a set of breakpoint widths, then declaratively render component
+sub-trees when those breakpoints are met. It looks something like this:
+
+```xml
+<>
+  <Media at="xs">
+    <MobileLayout />
+  </Media>
+  <Media greaterThan="xs">
+    <NonMobileLayout />
+  </Media>
+</>
+```
+
+In this example, we're emitting the `MobileLayout` component for devices at or below our `xs` breakpoint, and the
+`NonMobileLayout` for devices greater than our `xs` breakpoint. You can imagine that the `MobileLayout` and
+`NonMobileLayout` components are complicated sub-trees, with more significant differences than `styled-system`
+could handle.
+
+### How it works
+
+The first important thing to note is that when server-rendering with `react-responsive-media`, **all** breakpoints
+get rendered by the server.
+
+Why not just the one that the current device needs? A couple reasons. First, we can't _accurately_ identify which
+breakpoint your device needs from the server. We could use a library to sniff the browser `user-agent`, but those
+aren't always accurate, and they wouldn't give us all the information we need to know when we are server-rendering.
+
+The other reason `react-responsive-media` renders _all_ breakpoints has to do with the hydration of a
+server-rendered React app. When using SSR in a React app, you need to call
+[`ReactDOM.hydrate()`](https://reactjs.org/docs/react-dom.html#hydrate) to connect your server-rendered app to the
+client's browser. As the docs describe,
+
+> React expects that the rendered content is identical between the server and the client.
+
+If there are differences between your server-rendered component tree and the initial client-rendered component
+tree, a couple things will happen.
+
+1. You'll get a friendly error in your console telling you there were differences (in development mode), but
+   also...
+2. weird, unpredictable things will happen to your app.
+
+Here's what our app looked like in one case where our client render didn't match our server render:
+
+(TODO: image of spinning body)
+
+Yuck! The docs explain that `ReactDOM.hydrate` doesn't put a lot of effort into patching up differences between the
+server and client renders, and that's how you end up with weird things like above.
+
+Because hydration is so picky about the component trees matching between server and client, we render _all_
+breakpoints from the server with `react-responsive-media`. This way, we know you're going to get a branch that
+matches between the client and server. Then, upon hydration, we prune the branches that don't match the current
+breakpoint.
```

Comment:
Have we tested whether a warning gets thrown in the example? Once markup is passed down from the server and React attaches, I (believe, but could be wrong) it takes one render pass to prune, which would invariably trigger a hydration warning.

Reply:
* A warning gets thrown in _which_ example? The hypothetical being discussed in this section, or a literal example from above?
* Is your concern that "we should also make it clear that you'll still get a hydration mismatch warning", or that "something in this section is inaccurate"?

Reply:
> in the example

Meaning the example app in the RRM directory. My worry is that even with pruning we might trigger a hydration warning, which could be off-puting for folks. 

I'll pull it down later today and play with it a bit while I'm creating a few more example apps.

Reply:
I was able to verify that rehydration errors do get thrown in the example: 

<img width="891" alt="Screen Shot 2019-05-20 at 9 02 49 PM" src="https://user-images.githubusercontent.com/236943/58067584-afdc0200-7b42-11e9-948e-d8b6b5279bad.png">

Additionally, re
> Because hydration requires the component trees to match between server and client, we render _all_ breakpoints from the server with `react-responsive-media`. This way, we know you're going to get a branch that matches between the client and server.

@pepopowitz I brought this up [in another comment](https://github.com/artsy/artsy.github.io/pull/566#discussion_r283509489) (sorry for beating a dead horse here) but this is key to understanding the lib. Tying react's hydration requirements to the emission of html layout at each breakpoint mischaracterizes the purpose of this process and the novelty of the idea and we should avoid confusion here. Rendering out all breakpoints allows us a pre-JS-boot jumpstart for matching correct layout, via CSS, since CSS can execute without JS. Immediately, once content is sent down the wire to the browser, we get the correct layout  _without need for a user agent_ matching mobile, or `md` or whatever outlandish media query one might need because we don't need JS to determine whether the media query matches or not. Once react attaches, it simply washes over what is unneeded, via a JS-based `matchMedia` call -- also unrelated to React and its hydration characteristics. 

@alloy - would you mind clarifying as well as commenting on the hydration errors in the example? I don't see a way around them and consider it a good case for `suppressHydrationWarning`.

Reply:
@damassi I'm having trouble recognizing a significant difference between what I wrote and what you're saying. Would you mind suggesting specific changes to the article that would accomplish the point you're making?

Reply:
> Rendering out all breakpoints allows us a pre-JS-boot jumpstart for matching correct layout, via CSS, since CSS can execute without JS. 

I’ve tried to address this with [this here comment](https://github.com/artsy/artsy.github.io/pull/566#discussion_r286085247).

Reply:
> @alloy - would you mind clarifying as well as commenting on the hydration errors in the example? I don't see a way around them and consider it a potential case for suppressHydrationWarning.

@damassi Correct. I asked @kierangillen to make a PR for that, but I guess with his travelling he didn’t get around to that yet. I think we should make that fix for v1 and the launch of this post.

Reply:
@alloy will do 👍And ty, your comment concisely summarizes what I'm trying to say.

@pepopowitz this part: 

> Because hydration requires the component trees to match between server and client, we render all breakpoints from the server with react-responsive-media. This way, we know you're going to get a branch that matches between the client and server.

I would suggest removing this paragraph completely. The "because hydration requires component trees to match" and "we render out all breakpoints" creates a false correlation as this has nothing to do with React or hydration. Willing to explain more offline if need be.